lib/src/cache.rs

```diff
@@ -159,6 +164,24 @@ impl CacheEntry {
     pub fn take_go_get(&mut self) -> Option<Obligation> {
         self.go_get.take()
     }
+
+    /// Insert the provided body into the cache.
+    pub fn insert(&mut self, options: WriteOptions, body: Body) -> Result<(), crate::Error> {
+        let go_get = self.take_go_get().ok_or(Error::NotRevalidatable)?;
+        go_get.insert(options, body);
+        Ok(())
+    }
+
+    /// Freshen the existing body according to the provided options.
+    pub fn update(&mut self, options: WriteOptions) -> Result<(), crate::Error> {
+        let go_get = self.take_go_get().ok_or(Error::NotRevalidatable)?;
+        let Err((go_get, err)) = go_get.update(options) else {
+            return Ok(());
+        };
+        // On failure, preserve the obligation.
+        self.go_get = Some(go_get);
+        Err(err)
```

Comment:
Style nit: I think this would be easier to read with a `match`, despite the extra indent level. Not a blocker.